### ISSUE TITLE
presigning an endpoint with port doesn't work

### ISSUE THREAD

**Issue by vlovich** [aws/aws-sdk-js-v3#3858]
### Checkboxes for prior research

- [X] I've gone through [Developer Guide](https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide) and [API reference](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest)
- [X] I've checked [AWS Forums](https://forums.aws.amazon.com) and [StackOverflow](https://stackoverflow.com/questions/tagged/aws-sdk-js).
- [X] I've searched for [previous similar issues](https://github.com/aws/aws-sdk-js-v3/issues) and didn't find any solution.

### Describe the bug

There was a previously filed bug https://github.com/aws/aws-sdk-js-v3/issues/2726. The fix was attempted here: https://github.com/aws/aws-sdk-js-v3/pull/2773. It seems like the bug still exists because having a non-default port number still causes the SDK to generate the wrong signature.

### SDK version number

@aws-sdk/s3-request-presigner@3.145.0

### Which JavaScript Runtime is this issue in?

Node.js

### Details of the browser/Node.js/ReactNative version

v16.11.1

### Reproduction Steps

```js
import * as S3 from '@aws-sdk/client-s3'
import { getSignedUrl } from '@aws-sdk/s3-request-presigner'

const s3 = new S3.S3({
  region: 'us-east-1', // <-- needs to be set correctly if actually fetching
  credentials: {
    accessKeyId: '...', // <-- needs to be set correctly if actually fetching
    secretAcessKey: '...', // <-- needs to be set correctly if actually fetching
  },
  endpoint: 'https://127.0.0.1:12345', // <-- needs to be set correctly if actually fetching
  forcePathStyle: true,
})
const url = getSignedUrl(s3, new S3.GetObjectCommand({ Bucket: 'bucket', Key: 'key' })
await fetch(url)
```

### Observed Behavior

The signature is generated WITHOUT the port number included.

### Expected Behavior

The signature is generated WITH the port number included.

### Possible Solution

The previous fix was:
```
    if (!requestToSign.headers["host"]) {
      requestToSign.headers.host = requestToSign.hostname;
+      if (requestToSign.port) {
+       requestToSign.headers.host = `${requestToSign.headers.host}:${requestToSign.port}`;
+      }
    }
```
I think the correct fix is:
```
    if (!requestToSign.headers["host"]) {
      requestToSign.headers.host = requestToSign.hostname;
    }
+      if (requestToSign.port) {
+       requestToSign.headers.host = `${requestToSign.headers.host}:${requestToSign.port}`;
+      }
```
In other words the port needs to be appended even if "host" appears in the header. The unit test added in that PR unfortunately doesn't catch this issue because it missed that the middleware `@aws-sdk/middleware-host-header` injects the `host` header unconditionally AFAICT (or at least it does today).

**Comment by ajredniwja** [comment]
Hey @vlovich thanks for reporting this, you can open a PR for the team to review or I can take it up soon.

**Comment by KnownRock** [comment]
For everyone who get here, this is a way to bypass this bug.
```js
import { SignatureV4 } from '@aws-sdk/signature-v4'
import { Sha256 } from '@aws-crypto/sha256-browser'
import { HttpRequest } from '@aws-sdk/types'
import {
  S3Client,
  ListBucketsCommand
} from '@aws-sdk/client-s3'
const s3 = new S3Client({
  region: "us-east-1",
  credentials: {
    accessKeyId: 'xxxxxx',
    secretAccessKey: 'xxxxxx',
  },
  endpoint: "http://127.0.0.1:9000",
  forcePathStyle: true,
  signer: async () => ({
    sign: async (request: HttpRequest) => {
      request.headers['host'] = `${request.hostname}:${request.port}`

      const signatureV4 = new SignatureV4({
        credentials: {
          accessKeyId: 'xxxxxx',
          secretAccessKey: 'xxxxxx',
        },
        region: 'us-east-1',
        service: 's3',
        sha256: Sha256,
      });

      const authorizatedRequest = await signatureV4.sign(request);

      return authorizatedRequest
    }
  })
});


s3.send(new ListBucketsCommand({})).then((data) => {
  console.log(data)
})
```

**Comment by oscarhermoso** [comment]
@KnownRock - Beautiful! I replaced `@aws-sdk/signature-v4` with `@smithy/signature-v4` because the former is deprecated and it worked perfectly.

```diff
- import { SignatureV4 } from '@aws-sdk/signature-v4'
+ import { SignatureV4 } from '@smithy/signature-v4';
```

**Comment by KnownRock** [comment]
> @KnownRock - Beautiful! I replaced `@aws-sdk/signature-v4` with `@smithy/signature-v4` because the former is deprecated and it worked perfectly.
> 
> ```diff
> - import { SignatureV4 } from '@aws-sdk/signature-v4'
> + import { SignatureV4 } from '@smithy/signature-v4';
> ```

Thank you for the correction to the dependencies, I'll change the above example😉
